hledger-web/hledger-web.1

.TH "HLEDGER\-WEB" "1" "August 2026" "hledger-web-1.99 " "hledger User Manuals"



.SH NAME
hledger\-web \- web interface and API for \f[CR]hledger\f[R], a robust,
friendly plain text accounting app.
.SH SYNOPSIS
\f[CR]hledger\-web [OPTS] [QUERY]\f[R]
.PD 0
.P
.PD
or
.PD 0
.P
.PD
\f[CR]hledger web [OPTS] [QUERY]\f[R]
.SH DESCRIPTION
This manual is for hledger\(aqs web interface, version 1.99.
See also the hledger manual for common concepts and file formats.
.PP
hledger is a robust, user\-friendly, cross\-platform set of programs for
tracking money, time, or any other commodity, using double\-entry
accounting and a simple, editable file format.
hledger is inspired by and largely compatible with ledger(1), and
largely interconvertible with beancount(1).
.PP
hledger\-web is a simple web application for browsing and adding
transactions.
It provides a more user\-friendly UI than the hledger CLI or hledger\-ui
TUI, showing more at once (accounts, the current account register,
balance charts) and allowing history\-aware data entry, interactive
searching, and bookmarking.
.PP
hledger\-web also lets you share a journal with multiple users, or even
the public web.
There is no access control, so if you need that you should put it behind
a suitable web proxy.
As a small protection against data loss when running an unprotected
instance, it writes a numbered backup of the main journal file (only) on
every edit.
.PP
Like hledger, it reads from (and appends to) a journal file specified by
the \f[CR]LEDGER_FILE\f[R] environment variable (defaulting to
\f[CR]$HOME/.hledger.journal\f[R]); or you can specify files with
\f[CR]\-f\f[R] options.
It can also read timeclock files, timedot files, or any CSV/SSV/TSV file
with a date field.
(See hledger(1) \-> Input for details.)
.PP
hledger\-web can be run in three modes:
.IP \(bu 2
\f[CR]\-\-serve\-browse\f[R] mode (the default): the app serves the web
UI and JSON API, and opens your default web browser to show the app if
possible, and exits automatically after two minutes of inactivity (with
no requests received and no open browser windows viewing it).
.IP \(bu 2
\f[CR]\-\-serve\f[R]: the app just serves the web UI and JSON API.
.IP \(bu 2
\f[CR]\-\-serve\-api\f[R]: the app just serves the JSON API.
.PP
In all cases hledger\-web runs as a foreground process, logging requests
to stdout.
.SH OPTIONS
hledger\-web provides the following options:
.IP
.EX
Flags:
     \-\-serve \-\-server       serve and log requests, don\(aqt browse or auto\-exit
     \-\-serve\-api            like \-\-serve, but serve only the JSON web API,
                            not the web UI
     \-\-allow=view|add|edit  set the user\(aqs access level for changing data
                            (default: \(gaadd\(ga on a local\-only address, \(gaview\(ga
                            otherwise). It also accepts \(gasandstorm\(ga for that
                            platform (reads from the \(gaX\-Sandstorm\-Permissions\(ga
                            request header).
     \-\-cors=ORIGIN          allow cross\-origin requests from the specified
                            origin; setting ORIGIN to \(dq*\(dq allows requests from
                            any origin
     \-\-host=IPADDR          listen on this IP address (default: 127.0.0.1)
     \-\-port=PORT            listen on this TCP port (default: 5000)
     \-\-socket=SOCKET        listen on the given unix socket instead of an IP
                            address and port (unix only; implies \-\-serve)
     \-\-base\-url=BASEURL     set the base url (default: http://IPADDR:PORT)
     \-\-test                 run hledger\-web\(aqs tests and exit. hspec test
                            runner args may follow a \-\-, eg: hledger\-web \-\-test
                            \-\-help
.EE
.PP
By default hledger\-web listens only on IP address \f[CR]127.0.0.1\f[R],
which be accessed only from the local machine.
.PP
To allow access from elsewhere, use \f[CR]\-\-host\f[R] to specify an
externally accessible address configured on this machine, The special
address \f[CR]0.0.0.0\f[R] causes it to listen on all of this
machine\(aqs addresses.
.PP
Similarly, you can use \f[CR]\-\-port\f[R] to listen on a TCP port other
than 5000.
This is useful if you want to run multiple hledger\-web instances on a
machine.
.PP
When \f[CR]\-\-socket\f[R] is used, hledger\-web creates and
communicates via a socket file instead of a TCP port.
This can be more secure, respects unix file permissions, and makes
certain use cases easier, such as running per\-user instances behind an
nginx reverse proxy.
(Eg:
\f[CR]proxy_pass http://unix:/tmp/hledger/${remote_user}.socket;\f[R].)
.PP
You can use \f[CR]\-\-base\-url\f[R] to change the protocol, hostname,
port and path that appear in hledger\-web\(aqs hyperlinks.
This is useful eg when integrating hledger\-web within a larger website.
The default is \f[CR]http://HOST:PORT/\f[R] using the server\(aqs
configured host address and TCP port (or \f[CR]http://HOST\f[R] if PORT
is 80).
Note this affects url generation but not route parsing.
.PP
hledger\-web also supports many of hledger\(aqs general options:
.IP
.EX
General input flags:
  \-f \-\-file=[FMT:]FILE      Use this as the journal file (\- means stdin). If
                            not specified, $LEDGER_FILE or \(ti/.hledger.journal
                            will be used. If specified more than once, the
                            files will be read in order. Each file\(aqs format
                            (journal, csv, timeclock, timedot, rules..) is
                            inferred from the file extension or a FMT: prefix.
                            Some commands (add, import) write to the (first)
                            file, and expect it to be in journal format.
     \-\-rules=RULESFILE      Use rules defined in this rules file for
                            converting subsequent CSV/SSV/TSV files. If not
                            specified, uses FILE.csv.rules for each FILE.csv.
     \-\-alias=A=B|/RGX/=RPL  transform account names from A to B, or by
                            replacing regular expression matches
     \-\-auto                 generate extra postings by applying auto posting
                            rules (\(dq=\(dq) to all transactions
     \-\-forecast[=PERIOD]    Generate extra transactions from periodic rules
                            (\(dq\(ti\(dq), from after the latest ordinary transaction
                            until 6 months from now. Or, during the specified
                            PERIOD (the equals is required). Auto posting rules
                            will also be applied to these transactions. In
                            hledger\-ui, also make future\-dated transactions
                            visible at startup.
     \-\-ignore\-assertions    don\(aqt check balance assertions by default
     \-\-ignore\-lots          don\(aqt check lot entries by default
  \-I                        shortcut for \-\-ignore\-assertions \-\-ignore\-lots
     \-\-txn\-balancing=...    how to check that transactions are balanced:
                            \(aqold\(aq:   \- use global display precision
                            \(aqexact\(aq: \- use transaction precision (default)
     \-\-infer\-costs          infer conversion equity postings from costs
     \-\-infer\-equity         infer costs from conversion equity postings
     \-\-infer\-market\-prices  infer market prices from costs
     \-\-pivot=TAGNAME        use a different field or tag as account names
  \-s \-\-strict               do extra error checks (and override \-I)

General output flags (affecting some commands):
  \-b \-\-begin=DATE           include postings/transactions on/after this date
  \-e \-\-end=DATE             include postings/transactions before this date
                            (with a report interval, will be adjusted to
                            following subperiod end)
  \-D \-\-daily                set report interval: 1 day
  \-W \-\-weekly               set report interval: 1 week
  \-M \-\-monthly              set report interval: 1 month
  \-Q \-\-quarterly            set report interval: 1 quarter
  \-Y \-\-yearly               set report interval: 1 year
  \-p \-\-period=PERIODEXP     set begin date, end date, and/or report interval,
                            with more flexibility
     \-\-today=DATE           override today\(aqs date (affects relative dates)
     \-\-date2                match/use secondary dates instead (deprecated)
  \-U \-\-unmarked             include only unmarked postings/transactions
  \-P \-\-pending              include only pending postings/transactions
  \-C \-\-cleared              include only cleared postings/transactions
                            (\-U/\-P/\-C can be combined)
  \-R \-\-real                 include only non\-virtual postings
  \-E \-\-empty                Show zero items, which are normally hidden.
                            In hledger\-ui & hledger\-web, do the opposite.
     \-\-depth=DEPTHEXP       if a number (or \-NUM): show only top NUM levels
                            of accounts. If REGEXP=NUM, only apply limiting to
                            accounts matching the regular expression.
  \-B \-\-cost                 convert amounts to their cost/sale amount (\(at/\(at\(at)
  \-V \-\-market               valuation mode: show amounts converted to market
                            value at period end(s) in their default valuation
                            commodity. Short for \-\-value=end.
  \-X \-\-exchange=COMM        valuation mode: show amounts converted to market
                            value at period end(s) in the specified commodity.
                            Short for \-\-value=end,COMM.
     \-\-value=WHEN[,COMM]    valuation mode: show amounts converted to market
                            value on the specified date(s) in their default
                            valuation commodity or a specified commodity. WHEN
                            can be:
                            \(aqthen\(aq:     value on transaction dates
                            \(aqend\(aq:      value at period end(s)
                            \(aqnow\(aq:      value today
                            YYYY\-MM\-DD: value on given date
     \-\-lots                 show lot subaccounts and other lot details
  \-c \-\-commodity\-style=S    Override a commodity\(aqs display style.
                            Eg: \-c \(aq.\(aq or \-c \(aq1.000,00 EUR\(aq
     \-\-pretty[=YN]          Use box\-drawing characters in text output? The
                            optional \(aqy\(aq/\(aqyes\(aq or \(aqn\(aq/\(aqno\(aq arg requires =.
     \-\-title=T              set or customise a report title
     \-\-subreport\-titles=TS  customise subreport headings in compound reports
                            (|\-separated)
     \-\-period\-titles=OPT    customise headings in periodic reports:
                            \(aqcompact\(aq: readable period names when possible
                            \(aqdates\(aq:   exact dates/date ranges always

General help flags:
  \-h \-\-help                 show command line help
     \-\-tldr                 show command examples with tldr
     \-\-info                 show the manual with info
     \-\-man                  show the manual with man
     \-\-version              show version information
     \-\-debug=[1\-9]          show this much debug output (default: 1)
     \-\-pager=YN             use a pager when needed ? y/yes (default) or n/no
     \-\-color=YNA \-\-colour   use ANSI color ? y/yes, n/no, or auto (default)
.EE
.PP
hledger\-web shows accounts with zero balances by default (like
\f[CR]hledger\-ui\f[R], and unlike \f[CR]hledger\f[R]).
Using the \f[CR]\-E/\-\-empty\f[R] flag will reverse this behaviour.
If you see accounts which appear to have a zero balance, but cannot be
hidden with \f[CR]\-E\f[R], it\(aqs because they have a mixed\-cost
balance, which looks like zero when costs are hidden.
(hledger\-web does not show costs.)
.PP
Reporting options and/or query arguments can be used to set an initial
query, which although not shown in the UI, will restrict the data shown
(in addition to any search query entered in the UI).
.PP
If you use the bash shell, you can auto\-complete flags by pressing TAB
in the command line.
If this is not working see Install > Shell completions.
.SH PERMISSIONS
You can restrict who can access hledger\-web by
.IP \(bu 2
setting the IP address it listens on (see \f[CR]\-\-host\f[R] above).
By default it listens on 127.0.0.1, accessible to users on the local
machine only.
.IP \(bu 2
or by putting it behind an authenticating proxy, such as caddy or apache
.IP \(bu 2
or by putting it behind a firewall.
.PP
And you can restrict what the users reaching it can do, by specifying
the \f[CR]\-\-allow=ACCESSLEVEL\f[R] option at startup.
ACCESSLEVEL is one of:
.IP \(bu 2
\f[CR]view\f[R] \- allows viewing the journal file(s)
.IP \(bu 2
\f[CR]add\f[R] \- also allows adding new transactions to the main
journal file
.IP \(bu 2
\f[CR]edit\f[R] \- also allows editing, uploading or downloading the
journal file(s)
.IP \(bu 2
\f[CR]sandstorm\f[R] \- (for the hledger\-web Sandstorm app:) allows
whichever of \f[CR]view\f[R], \f[CR]add\f[R], or \f[CR]edit\f[R] are
specified in the \f[CR]X\-Sandstorm\-Permissions\f[R] HTTP header
.PP
The default access level is \f[CR]add\f[R] when listening on a
local\-only address (\f[CR]127.0.0.1\f[R], \f[CR]::1\f[R],
\f[CR]localhost\f[R], or a unix socket), and \f[CR]view\f[R] otherwise.
To allow more than the default access, start it with an explicit
\f[CR]\-\-allow=add\f[R] or \f[CR]\-\-allow=edit\f[R] option.
.SH EDITING, UPLOADING, DOWNLOADING
If you enable the \f[CR]manage\f[R] capability mentioned above,
you\(aqll see a new \(dqspanner\(dq button to the right of the search
form.
Clicking this will let you edit, upload, or download the journal file or
any files it includes.
.PP
Note, unlike any other hledger command, in this mode you (or any
visitor) can alter or wipe the data files.
.PP
Normally whenever a file is changed in this way, hledger\-web saves a
numbered backup (assuming file permissions allow it, the disk is not
full, etc.)
hledger\-web is not aware of version control systems, currently; if you
use one, you\(aqll have to arrange to commit the changes yourself (eg
with a cron job or a file watcher like entr).
.PP
Changes which would leave the journal file(s) unparseable or non\-valid
(eg with failing balance assertions) are prevented.
(Probably.
This needs re\-testing.)
.SH RELOADING
hledger\-web detects changes made to the files by other means (eg if you
edit it directly, outside of hledger\-web), and it will show the new
data when you reload the page or navigate to a new page.
If a change makes a file unparseable, hledger\-web will display an error
message until the file has been fixed.
.PP
(Note: if you are viewing files mounted from another machine, make sure
that both machine clocks are roughly in step.)
.SH JSON API
In addition to the web UI, hledger\-web also serves a JSON API that can
be used to get data or add new transactions.
If you want the JSON API only, you can use the \f[CR]\-\-serve\-api\f[R]
flag.
Eg:
.IP
.EX
$ hledger\-web \-f examples/sample.journal \-\-serve\-api
\&...
.EE
.PP
You can get JSON data from these routes:
.IP
.EX
/version
/accountnames
/transactions
/prices
/commodities
/accounts
/accounttransactions/ACCOUNTNAME
.EE
.PP
Eg, all account names in the journal (similar to the accounts command).
(hledger\-web\(aqs JSON does not include newlines, here we use python to
prettify it):
.IP
.EX
$ curl \-s http://127.0.0.1:5000/accountnames | python \-m json.tool
[
    \(dqassets\(dq,
    \(dqassets:bank\(dq,
    \(dqassets:bank:checking\(dq,
    \(dqassets:bank:saving\(dq,
    \(dqassets:cash\(dq,
    \(dqexpenses\(dq,
    \(dqexpenses:food\(dq,
    \(dqexpenses:supplies\(dq,
    \(dqincome\(dq,
    \(dqincome:gifts\(dq,
    \(dqincome:salary\(dq,
    \(dqliabilities\(dq,
    \(dqliabilities:debts\(dq
]
.EE
.PP
Or all transactions:
.IP
.EX
$ curl \-s http://127.0.0.1:5000/transactions | python \-m json.tool
[
    {
        \(dqtcode\(dq: \(dq\(dq,
        \(dqtcomment\(dq: \(dq\(dq,
        \(dqtdate\(dq: \(dq2008\-01\-01\(dq,
        \(dqtdate2\(dq: null,
        \(dqtdescription\(dq: \(dqincome\(dq,
        \(dqtindex\(dq: 1,
        \(dqtpostings\(dq: [
            {
                \(dqpaccount\(dq: \(dqassets:bank:checking\(dq,
                \(dqpamount\(dq: [
                    {
                        \(dqacommodity\(dq: \(dq$\(dq,
                        \(dqaismultiplier\(dq: false,
                        \(dqaprice\(dq: null,
\&...
.EE
.PP
Most of the JSON corresponds to hledger\(aqs data types; for details of
what the fields mean, see the Hledger.Data.Json haddock docs and click
on the various data types, eg Transaction.
And for a higher level understanding, see the journal docs.
There is also a basic OpenAPI specification.
.PP
In some cases there is outer JSON corresponding to a \(dqReport\(dq
type.
To understand that, go to the Hledger.Web.Handler.MiscR haddock and look
at the source for the appropriate handler to see what it returns.
Eg for \f[CR]/accounttransactions\f[R] it\(aqs getAccounttransactionsR,
returning a \(dq\f[CR]accountTransactionsReport ...\f[R]\(dq.
Looking up the haddock for that we can see that /accounttransactions
returns an AccountTransactionsReport, which consists of a report title
and a list of AccountTransactionsReportItem (etc).
.PP
You can add a new transaction to the journal with a PUT request to
\f[CR]/add\f[R], if hledger\-web was started with the \f[CR]add\f[R]
capability (enabled by default).
The payload must be the full, exact JSON representation of a hledger
transaction (partial data won\(aqt do).
You can get sample JSON from hledger\-web\(aqs \f[CR]/transactions\f[R]
or \f[CR]/accounttransactions\f[R], or you can export it with
hledger\-lib, eg like so:
.IP
.EX
\&.../hledger$ stack ghci hledger\-lib
>>> writeJsonFile \(dqtxn.json\(dq (head $ jtxns samplejournal)
>>> :q
.EE
.PP
Here\(aqs how it looks as of hledger\-1.17 (remember, this JSON
corresponds to hledger\(aqs Transaction and related data types):
.IP
.EX
{
    \(dqtcomment\(dq: \(dq\(dq,
    \(dqtpostings\(dq: [
        {
            \(dqpbalanceassertion\(dq: \f[B]null\f[R],
            \(dqpstatus\(dq: \(dqUnmarked\(dq,
            \(dqpamount\(dq: [
                {
                    \(dqaprice\(dq: \f[B]null\f[R],
                    \(dqacommodity\(dq: \(dq$\(dq,
                    \(dqaquantity\(dq: {
                        \(dqfloatingPoint\(dq: 1,
                        \(dqdecimalPlaces\(dq: 10,
                        \(dqdecimalMantissa\(dq: 10000000000
                    },
                    \(dqaismultiplier\(dq: \f[B]false\f[R],
                    \(dqastyle\(dq: {
                        \(dqascommodityside\(dq: \(dqL\(dq,
                        \(dqasdigitgroups\(dq: \f[B]null\f[R],
                        \(dqascommodityspaced\(dq: \f[B]false\f[R],
                        \(dqasprecision\(dq: 2,
                        \(dqasdecimalpoint\(dq: \(dq.\(dq
                    }
                }
            ],
            \(dqptransaction_\(dq: \(dq1\(dq,
            \(dqpaccount\(dq: \(dqassets:bank:checking\(dq,
            \(dqpdate\(dq: \f[B]null\f[R],
            \(dqptype\(dq: \(dqRegularPosting\(dq,
            \(dqpcomment\(dq: \(dq\(dq,
            \(dqpdate2\(dq: \f[B]null\f[R],
            \(dqptags\(dq: [],
            \(dqporiginal\(dq: \f[B]null\f[R]
        },
        {
            \(dqpbalanceassertion\(dq: \f[B]null\f[R],
            \(dqpstatus\(dq: \(dqUnmarked\(dq,
            \(dqpamount\(dq: [
                {
                    \(dqaprice\(dq: \f[B]null\f[R],
                    \(dqacommodity\(dq: \(dq$\(dq,
                    \(dqaquantity\(dq: {
                        \(dqfloatingPoint\(dq: \-1,
                        \(dqdecimalPlaces\(dq: 10,
                        \(dqdecimalMantissa\(dq: \-10000000000
                    },
                    \(dqaismultiplier\(dq: \f[B]false\f[R],
                    \(dqastyle\(dq: {
                        \(dqascommodityside\(dq: \(dqL\(dq,
                        \(dqasdigitgroups\(dq: \f[B]null\f[R],
                        \(dqascommodityspaced\(dq: \f[B]false\f[R],
                        \(dqasprecision\(dq: 2,
                        \(dqasdecimalpoint\(dq: \(dq.\(dq
                    }
                }
            ],
            \(dqptransaction_\(dq: \(dq1\(dq,
            \(dqpaccount\(dq: \(dqincome:salary\(dq,
            \(dqpdate\(dq: \f[B]null\f[R],
            \(dqptype\(dq: \(dqRegularPosting\(dq,
            \(dqpcomment\(dq: \(dq\(dq,
            \(dqpdate2\(dq: \f[B]null\f[R],
            \(dqptags\(dq: [],
            \(dqporiginal\(dq: \f[B]null\f[R]
        }
    ],
    \(dqttags\(dq: [],
    \(dqtsourcepos\(dq: {
        \(dqtag\(dq: \(dqJournalSourcePos\(dq,
        \(dqcontents\(dq: [
            \(dq\(dq,
            [
                1,
                1
            ]
        ]
    },
    \(dqtdate\(dq: \(dq2008\-01\-01\(dq,
    \(dqtcode\(dq: \(dq\(dq,
    \(dqtindex\(dq: 1,
    \(dqtprecedingcomment\(dq: \(dq\(dq,
    \(dqtdate2\(dq: \f[B]null\f[R],
    \(dqtdescription\(dq: \(dqincome\(dq,
    \(dqtstatus\(dq: \(dqUnmarked\(dq
}
.EE
.PP
And here\(aqs how to test adding it with curl.
This should add a new entry to your journal:
.IP
.EX
$ curl http://127.0.0.1:5000/add \-X PUT \-H \(aqContent\-Type: application/json\(aq \-\-data\-binary \(attxn.json
.EE
.SH DEBUG OUTPUT
You can add \f[CR]\-\-debug[=N]\f[R] to the command line to log debug
output.
N ranges from 1 (least output, the default) to 9 (maximum output).
Typically you would start with 1 and increase until you are seeing
enough.
Debug output goes to stderr, interleaved with the requests logged on
stdout.
To capture debug output in a log file instead, you can usually redirect
stderr, eg:
.PD 0
.P
.PD
\f[CR]hledger\-web \-\-debug=3 2>hledger\-web.log\f[R].
.SH ENVIRONMENT
\f[B]LEDGER_FILE\f[R] The main journal file to use when not specified
with \f[CR]\-f/\-\-file\f[R].
Default: \f[CR]$HOME/.hledger.journal\f[R].
.SH BUGS
We welcome bug reports in the hledger issue tracker
(https://bugs.hledger.org), or on the hledger chat or mail list
(https://hledger.org/support).
.PP
Some known issues:
.PP
Does not work well on small screens, or in text\-mode browsers.


.SH AUTHORS
Simon Michael <simon@joyful.com> and contributors.
.br
See http://hledger.org/CREDITS.html

.SH COPYRIGHT
Copyright 2007-2023 Simon Michael and contributors.

.SH LICENSE
Released under GNU GPL v3 or later.

.SH SEE ALSO
hledger(1), hledger\-ui(1), hledger\-web(1), ledger(1)


NEWS
hledger news
============

2010/05/23 hledger 0.10
........................

  * fix too-loose testpack dependency, missing safe dependency
  * fix ghc 6.12 compatibility with -fweb
  * fix handling of non-ascii arguments with ghc 6.12
  * fix "0.8" in --version output
  * fix an occasional stack overflow error due to infinite recursion in Posting/Transaction equality tests
  * the -fwebhappstack build flag is gone for now, to avoid a cabal problem
  * parsing: if there is no description, don't require a space after the transaction date
  * parsing: balance balanced-virtual postings separately, allow them to have an implicit amount
  * parsing: timelog entries now generate balanced transactions, using virtual postings
  * parsing: simpler high-level parse error message
  * parsing: clearer bad date errors
  * add: fix wrongful program exit on bad dates
  * print: negative account patterns now exclude transactions containing any posting to a matched account
  * vty: rename the ui command to vty for consistency
  * vty: fix restricted account scope when backing up to top level
  * web: fix non-ascii handling with ghc 6.12
  * web: fix a bug possibly affecting reload-on-change
  * consolidate module namespace under Hledger, api cleanups

  Stats:
  44 days, 81 commits since last release.
  Now at 4904 lines of code including tests, 144 tests, 53% coverage.

2010/04/10 hledger 0.9
......................

  * ghc 6.12 support
  * split off hledger-lib package, containing core types & utils
  * parsing: ignore D, C, N, tag, end tag directives; we should now accept any ledger 2.6 file
  * parsing: allow numbers in commodities if double-quoted, like ledger
  * parsing: allow transactions with empty descriptions
  * parsing: show a better error for illegal month/day numbers in dates
  * parsing: don't ignore trailing junk in a smart date, eg in web add form
  * parsing: don't ignore unparsed text following an amount
  * parsing: @ was being treated as a currency symbol
  * add: fix precision handling in default amounts (#19)
  * add: elide last amount in added transactions
  * convert: keep original description by default, allow backreferences in replace pattern
  * convert: basic csv file checking, warn instead of dying when it looks wrong
  * convert: allow blank/comment lines at end of rules file
  * print: always show zero amounts as 0, hiding any commodity/decimal places/price, like ledger
  * register: fix bad layout with years < 1000
  * register: fix a Prelude.head error with reporting interval, --empty, and --depth
  * register: fix a regression, register should not show posting comments
  * register: with --empty, intervals should continue to ends of the specified period
  * stats: better output when last transaction is in the future
  * stats: show commodity symbols, account tree depth, reorder slightly
  * web: -fweb now builds with simpleserver; to get happstack, use -fwebhappstack instead
  * web: pre-fill the add form with today's date
  * web: help links, better search form wording
  * web: show a proper error for a bad date in add form (#17)
  * web: fix for unicode search form values
  * web: fix stack overflow caused by regexpr, and handle requests faster (#14)
  * web: look for more-generic browser executables
  * web: more robust browser starting (#6)
  * error message cleanups
  * more tests, refactoring, docs

  Stats:
  58 days, 2 contributors, 102 commits since last release.
  Now at 3983 lines of non-test code, 139 tests, 53% coverage.

2010/02/11 hledger 0.8
......................

  * parsing: in date=date2, use first date's year as a default for the second
  * add: ctrl-d doesn't work on windows, suggest ctrl-c instead
  * add: --no-new-accounts option disallows new accounts (Roman Cheplyaka)
  * add: re-use the previous transaction's date as default (Roman Cheplyaka)
  * add: a command-line argument now filters by account during history matching (Roman Cheplyaka)
  * chart: new command, generates balances pie chart (requires -fchart flag, gtk2hs) (Roman Cheplyaka, Simon Michael)
  * register: make reporting intervals honour a display expression (#18)
  * web: fix help link
  * web: use today as default when adding with a blank date
  * web: re-enable account/period fields, they seem to be fixed, along with file re-reading (#16)
  * web: get static files from the cabal data dir, or the current dir when using make (#13)
  * web: preserve encoding during add, assuming it's utf-8 (#15)
  * fix some non-utf8-aware file handling (#15)
  * filter ledger again for each command, not just once at program start
  * refactoring, clearer data types

  Stats:
  62 days, 2 contributors, 76 commits since last release.
  Now at 3464 lines of non-test code, 97 tests, 53% test coverage.

2009/12/11 hledger 0.7
........................

  * price history support (first cut):
    P directives now work, though differently from c++ ledger. Each
    posting amount takes its fixed unit price from the price history (or
    @) when available. This is simple and useful for things like foreign
    currency expenses (but not investment tracking). Like ledger, balance
    and register don't show amount prices any more, and don't separate
    differently-priced amounts. Unlike ledger, print shows all amount
    prices, and supports -B.
  * --effective option, will use transactions' effective dates if any
  * convert: new rules file format, find/create rules file automatically,
    more robust parsing, more useful --debug output
  * print: always sort by date, fix long account name truncation, align
    amounts, show end of line comments, show all amounts for clarity
    (don't elide the final balancing amount)
  * ui: use vty 4, fixes non-ascii and gnome terminal problems (issues #3, #4)
  * web: allow data entry, react to data file changes, better layout, help
    links, remove histogram command and filter fields for now, fix bad

hledger-web/hledger-web.hs
{-# LANGUAGE CPP, OverloadedStrings #-}
{-|

hledger-web - a hledger add-on providing a web interface.
Copyright (c) 2007-2012 Simon Michael <simon@joyful.com>
Released under GPL version 3 or later.

-}

module Main
where

import Data.Conduit.Network (HostPreference(..))
import Network.Wai.Handler.Warp (runSettings, defaultSettings, settingsPort)
import Yesod.Default.Config
-- import Yesod.Default.Main   (defaultMain)
import Yesod.Logger ({- Logger,-} defaultDevelopmentLogger) --, logString)

import Prelude hiding (putStrLn)
-- -- import Control.Concurrent (forkIO, threadDelay)
import Control.Monad
-- import Data.Maybe
import Data.Text(pack)
import System.Exit
import System.IO.Storage (withStore, putValue)
import Text.Printf

import Hledger
import Hledger.Cli hiding (progname,prognameandversion)
import Hledger.Utils.UTF8IOCompat (putStrLn)
import Hledger.Web hiding (opts,j)


main :: IO ()
main = do
  opts <- getHledgerWebOpts
  when (debug_ $ cliopts_ opts) $ printf "%s\n" prognameandversion >> printf "opts: %s\n" (show opts)
  runWith opts

runWith :: WebOpts -> IO ()
runWith opts
  | "help" `in_` (rawopts_ $ cliopts_ opts)            = putStr (showModeHelp webmode) >> exitSuccess
  | "version" `in_` (rawopts_ $ cliopts_ opts)         = putStrLn prognameandversion >> exitSuccess
  | "binary-filename" `in_` (rawopts_ $ cliopts_ opts) = putStrLn (binaryfilename progname)
  | otherwise                                          = journalFilePathFromOpts (cliopts_ opts) >>= requireJournalFileExists >> withJournalDo' opts web

withJournalDo' :: WebOpts -> (WebOpts -> Journal -> IO ()) -> IO ()
withJournalDo' opts cmd = do
  journalFilePathFromOpts (cliopts_ opts) >>= readJournalFile Nothing Nothing >>=
    either error' (cmd opts . journalApplyAliases (aliasesFromOpts $ cliopts_ opts))

-- | The web command.
web :: WebOpts -> Journal -> IO ()
web opts j = do
  -- unless (debug_ $ cliopts_ opts) $ forkIO (browser baseurl) >> return ()
  server (base_url_ opts) (port_ opts) opts j

-- browser :: String -> IO ()
-- browser baseurl = do
--   threadDelay $ fromIntegral browserstartdelay
--   putStrLn "Attempting to start a web browser"
--   openBrowserOn baseurl >> return ()

server :: String -> Int -> WebOpts -> Journal -> IO ()
server baseurl port opts j = do
  printf "Starting http server on port %d with base url %s\n" port baseurl
  -- let a = App{getStatic=static staticdir
  --            ,appRoot=pack baseurl
  --            ,appOpts=opts
  --            ,appArgs=patterns_ $ reportopts_ $ cliopts_ opts
  --            ,appJournal=j
  --            }
  withStore "hledger" $ do
    putValue "hledger" "journal" j

-- defaultMain :: (Show env, Read env)
--             => IO (AppConfig env extra)
--             -> (AppConfig env extra -> Logger -> IO Application)
--             -> IO ()
-- defaultMain load getApp = do
    -- config <- fromArgs parseExtra
    let config = AppConfig {
              appEnv = Development
            , appPort = port_ opts
            , appRoot = pack baseurl
            , appHost = HostIPv4
            , appExtra = Extra "" Nothing
            }
    logger <- defaultDevelopmentLogger
    app <- getApplication config logger
    runSettings defaultSettings
        { settingsPort = appPort config
        } app
